src/atc/etl/loaders/UpsertLoader.py

```diff
@@ -1,101 +1,23 @@
-
-        mode = "append" if incremental_load else "overwrite"
 
-        self.params.target_dh.write_or_append(df, mode=mode)
-        print(
-            "Incremental Base - incremental load with append"
-            if incremental_load
-            else "Incremental Base - full load with append"
-        )
-        return df
+        self.handle = handle
+        self.join_cols = join_cols
 
     def save_many(self, datasets: dataset_group) -> None:
-        pass
-
-    def save(self, df_source: DataFrame):
-        if df_source is None:
-            return df_source
-
-        df = df_source
-
-        # check null keys in our dataframe.
-        any_null_keys = len(
-            df.filter(
-                " OR ".join(f"({col} is NULL)" for col in self.params.join_cols)
-            ).take(1)
-        )
-
-        if any_null_keys:
-            print(
-                "Null keys found in input dataframe. "
-                "Rows will be discarded before load."
-            )
-            df = df.filter(
-                " AND ".join(f"({col} is NOT NULL)" for col in self.params.join_cols)
-            )
-
-        # Load data from the target table for the purpose of incremental load
-        if not self.params.incremental_load:
-            return self.do_simple_save(df)
-
-        df_target = self.params.target_dh.read()
-
-        # If the target is empty, always do faster full load
-        if len(df_target.take(1)) == 0:
-            return self.do_simple_save(df, incremental_load=False)
-
-        # Find records that need to be updated in the target (happens seldom)
-
-        # Define the column to be used for checking for new rows
-        # Checking the null-ness of one right row is sufficient to mark the row as new,
-        # since null keys are disallowed.
-
-        df, merge_required = CheckDfMerge(
-            df=df,
-            df_target=df_target,
-            join_cols=self.params.join_cols,
-            avoid_cols=[],
-        )
-
-        if not merge_required:
-            return self.do_simple_save(df)
-
-        temp_view_name = get_unique_tempview_name()
-        df.createOrReplaceTempView(temp_view_name)
-
-        target_table_name = TableConfigurator().table_name(self.params.target_id)
-        non_join_cols = [col for col in df.columns if col not in self.params.join_cols]
-
-        merge_sql_statement = GetMergeStatement(
-            merge_statement_type="delta",
-            target_table_name=target_table_name,
-            source_table_name=temp_view_name,
-            join_cols=self.params.join_cols,
-            insert_cols=df.columns,
-            update_cols=non_join_cols,
-            special_update_set="",
-        )
-
-        Spark.get().sql(merge_sql_statement)
-
-        print("Incremental Base - incremental load with merge")
+        """Upserts each dataset in the datasetgroup individually."""
+        for df in datasets.values():
```

Comment:
To me this is surprising behavior. If I want this behavior, I would inject a UnionTransformer before the loader. I suggest we remove the save_many functionality completely.

Reply:
I get your point Simon.

We only use the save_many in a single loader called OrchestratorLoader. The OrchestratorLoader is used to pass an output of one orchestrator as input to a second orchestrator. Here we have the save_many that pass the dataset into the execute function. It could we argued that this Loader could be its own class based step. Then i don't see a use for the save_many on a Loader.

Reply:
I will remove the save many method. Also since it is much less efficient to upsert each dataframe seperately instead of using the UnionTransformer first - then call save().